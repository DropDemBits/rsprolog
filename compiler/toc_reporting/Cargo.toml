[package]
name = "toc_reporting"
version = "0.1.0"
authors = ["DropDemBits <r3usrlnd@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
toc_span = { path = "../toc_span" }

drop_bomb = "0.1.5"
