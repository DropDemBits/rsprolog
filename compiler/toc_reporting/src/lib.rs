//! Common message reporting for all compiler libraries
use std::fmt;

use toc_span::TextRange;

/// Type of message reported.
/// Kinds are ordered by severity (from least severe, to most severe).
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum MessageKind {
    Warning,
    Error,
}

/// Type of annotation added to a message
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AnnotateKind {
    /// Information related to the main message.
    /// May be context specific.
    Note,
    /// More detailed related to the message.
    Info,
}

/// A reported message
#[derive(Debug)]
pub struct ReportMessage {
    kind: MessageKind,
    msg: String,
    range: TextRange,
    annotations: Vec<Annotation>,
}

impl ReportMessage {
    /// Gets the kind of message reported
    pub fn kind(&self) -> MessageKind {
        self.kind
    }

    /// Gets the reported message
    pub fn message(&self) -> &str {
        &self.msg
    }

    /// Gets the range of text the message covers
    pub fn text_range(&self) -> TextRange {
        self.range
    }

    /// Gets any associated annotations
    pub fn annotations(&self) -> &[Annotation] {
        &self.annotations
    }
}

impl fmt::Display for ReportMessage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let (start, end) = (u32::from(self.range.start()), u32::from(self.range.end()));
        let kind = match self.kind() {
            MessageKind::Warning => "warn",
            MessageKind::Error => "error",
        };

        write!(f, "{} at {}..{}: {}", kind, start, end, self.msg)?;

        // Report any annotations
        for annotation in &self.annotations {
            write!(f, "\n| {}", annotation)?;
        }

        Ok(())
    }
}

#[derive(Debug)]
pub struct Annotation {
    kind: AnnotateKind,
    msg: String,
    range: Option<TextRange>,
}

impl Annotation {
    /// Gets the kind of annotation reported
    pub fn kind(&self) -> AnnotateKind {
        self.kind
    }

    /// Gets the annotation's message
    pub fn message(&self) -> &str {
        &self.msg
    }

    /// Gets the range of text the message covers.
    /// If `None`, no range should be reported either
    pub fn text_range(&self) -> Option<TextRange> {
        self.range
    }
}

impl fmt::Display for Annotation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let kind = match self.kind() {
            AnnotateKind::Info => "info",
            AnnotateKind::Note => "note",
        };

        if let Some(range) = self.range {
            let (start, end) = (u32::from(range.start()), u32::from(range.end()));
            write!(f, "{} for {}..{}: {}", kind, start, end, self.msg)
        } else {
            write!(f, "{}: {}", kind, self.msg)
        }
    }
}

/// Sink for message reports
#[derive(Debug, Default)]
pub struct MessageSink {
    messages: Vec<ReportMessage>,
}

impl MessageSink {
    pub fn new() -> Self {
        Self { messages: vec![] }
    }

    /// Merges an existing set of messages into a new MessageSink
    pub fn with_messages(messages: Vec<ReportMessage>) -> Self {
        Self { messages }
    }

    /// Reports a message
    ///
    /// Does not add any annotations to the message
    pub fn report(&mut self, kind: MessageKind, message: &str, range: TextRange) {
        MessageBuilder::new(self, kind, message, range).finish();
    }

    /// Reports a detailed message
    ///
    /// Returns a builder for adding annotations
    pub fn report_detailed(
        &mut self,
        kind: MessageKind,
        message: &str,
        range: TextRange,
    ) -> MessageBuilder {
        MessageBuilder::new(self, kind, message, range)
    }

    /// Removes any subsequent messages that share the same text range
    pub fn dedup_shared_ranges(&mut self) {
        self.messages
            .dedup_by(|a, b| a.range == b.range && a.kind == b.kind)
    }

    /// Finishes reporting any messages, giving back the final message list
    pub fn finish(self) -> Vec<ReportMessage> {
        self.messages
    }
}

/// Builder for detailed messages
#[derive(Debug)]
pub struct MessageBuilder<'a> {
    drop_bomb: drop_bomb::DropBomb,
    reporter: &'a mut MessageSink,
    kind: MessageKind,
    message: String,
    range: TextRange,
    annotations: Vec<Annotation>,
}

impl<'a> MessageBuilder<'a> {
    pub fn new(
        reporter: &'a mut MessageSink,
        kind: MessageKind,
        message: &str,
        range: TextRange,
    ) -> Self {
        Self {
            drop_bomb: drop_bomb::DropBomb::new("Missing `finish()` for MessageBuilder"),
            reporter,
            kind,
            message: message.to_string(),
            range,
            annotations: vec![],
        }
    }

    pub fn with_annotation<R>(mut self, kind: AnnotateKind, message: &str, range: R) -> Self
    where
        R: Into<Option<TextRange>>,
    {
        self.annotations.push(Annotation {
            kind,
            msg: message.to_string(),
            range: range.into(),
        });

        self
    }

    pub fn with_note<R>(self, message: &str, range: R) -> Self
    where
        R: Into<Option<TextRange>>,
    {
        self.with_annotation(AnnotateKind::Note, message, range)
    }

    pub fn with_info<R>(self, message: &str, range: R) -> Self
    where
        R: Into<Option<TextRange>>,
    {
        self.with_annotation(AnnotateKind::Info, message, range)
    }

    pub fn finish(self) {
        let MessageBuilder {
            mut drop_bomb,
            reporter,
            kind,
            message,
            range,
            annotations,
        } = self;

        // Defuse bomb now
        drop_bomb.defuse();

        reporter.messages.push(ReportMessage {
            kind,
            msg: message,
            range,
            annotations,
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn report_message() {
        let mut sink = MessageSink::new();
        sink.report(
            MessageKind::Error,
            "an error message",
            TextRange::new(1.into(), 3.into()),
        );

        sink.report(
            MessageKind::Warning,
            "a warning message",
            TextRange::new(3.into(), 5.into()),
        );

        let msgs = sink.finish();
        assert_eq!(
            (msgs[0].message(), msgs[0].text_range()),
            ("an error message", TextRange::new(1.into(), 3.into()))
        );

        assert_eq!(
            (msgs[1].message(), msgs[1].text_range()),
            ("a warning message", TextRange::new(3.into(), 5.into()))
        );
    }
}
