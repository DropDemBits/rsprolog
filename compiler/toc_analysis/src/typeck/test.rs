//! Type check tests
use std::sync::Arc;

use toc_reporting::ReportMessage;
use unindent::unindent;

use crate::{const_eval::ConstEvalCtx, ty::TyCtx};

macro_rules! test_for_each_op {
    ($top_level_name:ident, [$(($op:literal, $sub_name:ident)),+ $(,)?] => $source:literal) => {
        ::paste::paste! {
                $(
                #[test]
                fn [<$top_level_name _ $sub_name>]() {
                    let source = format!($source, $op);
                    assert_typecheck(&source);
                }
            )+
        }
    };
}

#[track_caller]
fn assert_typecheck(source: &str) {
    insta::assert_snapshot!(insta::internals::AutoName, do_typecheck(source), source);
}

#[track_caller]
fn assert_named_typecheck(name: &str, source: &str) {
    insta::assert_snapshot!(name, do_typecheck(source), source);
}

fn do_typecheck(source: &str) -> String {
    let parsed = toc_parser::parse(&source);
    let mut unit_map = toc_hir::UnitMapBuilder::new();
    let hir_res = toc_hir_lowering::lower_ast(parsed.syntax(), &mut unit_map);
    let unit_map = Arc::new(unit_map.finish());

    let unit = unit_map.get_unit(hir_res.id);
    let const_eval_ctx = Arc::new(ConstEvalCtx::new(unit_map.clone()));
    let (ty_ctx, typeck_messages) = crate::typeck::typecheck_unit(unit, const_eval_ctx);

    stringify_typeck_results(&ty_ctx, &typeck_messages)
}

fn stringify_typeck_results(ty_ctx: &TyCtx, messages: &[ReportMessage]) -> String {
    let mut s = String::new();
    // Pretty print typectx
    let mut pretty_ty_ctx = crate::ty::pretty_dump_typectx(ty_ctx);
    // Trim trailing newline
    pretty_ty_ctx.pop();
    s.push_str(&pretty_ty_ctx);

    // Pretty print the messages
    for err in messages.iter() {
        s.push_str(&format!("\n{}", err));
    }

    s
}

#[test]
fn var_decl_type_spec() {
    assert_typecheck(r#"var k : string"#);
}

#[test]
fn var_decl_inference() {
    assert_typecheck(r#"var k := "oeuf""#);
}

#[test]
fn var_decl_init_typecheck() {
    assert_typecheck(r#"var k : string := "oeuf""#);
}

#[test]
fn bare_var_decl() {
    // Invariant, to be covered by the parser stage
    // Should not amount to anything
    assert_typecheck("var k");
}

#[test]
fn undeclared_symbol() {
    assert_typecheck("var a := b");
    // No cyclic deps
    assert_typecheck("var a := a");
}

// Typecheck numerical ops
test_for_each_op! {
    number_op,
    [
        ("+", add),
        ("-", sub),
        ("*", mul),
        ("div", idiv),
        ("/", rdiv),
        ("mod", r#mod),
        ("rem", rem),
        ("**", exp),
    ] => r#"
    % Compatiblitly with all variant of numbers
    var r : real
    var i : int
    var n : nat
    var _rr := r {0} r
    var _ri := r {0} i
    var _ir := i {0} r
    var _rn := r {0} n
    var _nr := n {0} r
    var _ii := i {0} i
    var _in := i {0} n
    var _ni := n {0} i
    var _nn := n {0} n
"#
}

// Test integer inference for all compatible operators
test_for_each_op!(
    integer_inferrence,
    [
        ("+", add),
        ("-", sub),
        ("*", mul),
        ("div", idiv),
        ("/", rdiv),
        ("mod", r#mod),
        ("rem", rem),
        ("**", exp),
    ] => r#"
    % Inferred integer types should pass
    % Decl should be a concrete type
    var a := 1 {0} 1
    % Types of operands should make the type concrete
    var r : real
    var i : int
    var n : nat
    var _r0 := 1 {0} r
    var _r1 := r {0} 1
    var _i0 := 1 {0} i
    var _i1 := i {0} 1
    var _n0 := 1 {0} n
    var _n1 := n {0} 1
"#);

#[test]
fn typecheck_error_prop() {
    // Only one error should be reported, propogated error supresses the rest
    assert_typecheck(&unindent(
        r#"
    var a : int
    var b : string
    var c := a + b
    var j := c + a
    "#,
    ));
}

#[test]
fn typecheck_sized_char() {
    assert_named_typecheck("sized_char_literal", r#"var _ : char(1)"#);
    // trip through negatives shouldn't affect anything
    assert_named_typecheck("sized_char_simple_expr", r#"var _ : char(1 - 1 * 1 + 2)"#);
    assert_named_typecheck("sized_char_zero_sized", r#"var _ : char(0)"#);
    assert_named_typecheck("sized_char_max_sized", r#"var _ : char(32768)"#);
    assert_named_typecheck("sized_char_wrong_type", r#"var _ : char(1.0)"#);
    assert_named_typecheck("sized_char_wrong_type_bool", r#"var _ : char(true)"#);
    assert_named_typecheck("sized_char_const_err", r#"var _ : char(1.0 div 0.0)"#);
}

#[test]
fn typecheck_sized_string() {
    assert_named_typecheck("sized_string_literal", r#"var _ : string(1)"#);
    // trip through negatives shouldn't affect anything
    assert_named_typecheck(
        "sized_string_simple_expr",
        r#"var _ : string(1 - 1 * 1 + 2)"#,
    );
    assert_named_typecheck("sized_string_zero_sized", r#"var _ : string(0)"#);
    assert_named_typecheck("sized_string_max_sized", r#"var _ : string(32768)"#);
    assert_named_typecheck("sized_string_wrong_type", r#"var _ : string(1.0)"#);
    assert_named_typecheck("sized_string_wrong_type_bool", r#"var _ : string(true)"#);
    assert_named_typecheck("sized_string_const_err", r#"var _ : string(1.0 div 0.0)"#);
}

#[test]
fn typeck_constvar_initializer() {
    assert_named_typecheck("typeck_var_compatible", r#"var k : int := 100"#);
    assert_named_typecheck("typeck_var_incompatible", r#"var k : char := 20"#);
    assert_named_typecheck(
        "typeck_var_error_prop",
        &unindent(
            r#"
    var k := 20 + false
    var l : int := k   % Nothing reported here
    "#,
        ),
    );

    assert_named_typecheck("typeck_const_compatible", r#"const k : int := 100"#);
    assert_named_typecheck("typeck_const_incompatible", r#"const k : char := 20"#);
    assert_named_typecheck(
        "typeck_const_error_prop",
        &unindent(
            r#"
    const k := 20 + false
    const l : int := k   % Nothing reported here
    "#,
        ),
    );
}
