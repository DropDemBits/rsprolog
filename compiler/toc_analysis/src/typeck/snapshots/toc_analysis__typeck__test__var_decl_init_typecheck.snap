---
source: compiler/toc_analysis/src/typeck/test.rs
expression: "var k : string := \"oeuf\""

---
ty_nodes:
    TypeId(HirId(Idx::<HirNode>(0))) TyRef(String)
def_kinds:
    DefId(0) Var(TyRef(String))
